gadget/update.go

```diff
@@ -148,8 +148,70 @@ func isCompatibleSchema(gadgetSchema, diskSchema string) bool {
 	}
 }
 
+func onDiskStructureIsLikelyImplicitSystemDataRole(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVolume, s OnDiskStructure) bool {
+	// in uc16/uc18 we used to allow system-data to be implicit / missing from
+	// the gadget.yaml in which case we won't have system-data in the laidOutVol
+	// but it will be in diskLayout, so we sometimes need to check if a given on
+	// disk partition looks like it was created implicitly by ubuntu-image as
+	// specified via the defaults in
+	// https://github.com/canonical/ubuntu-image-legacy/blob/master/ubuntu_image/parser.py#L568-L589
+
+	// namely it must meet the following conditions:
+	// * fs is ext4
+	// * partition type is "Linux filesystem data"
+	// * fs label is "writable"
+	// * this on disk structure is last on the disk
+	// * there is exactly one more structure on disk than partitions in the
+	//   gadget
+	// * the size of this structure extends to the end of the disk
+	// * there is no system-data role in the gadget.yaml
+
+	// bare structures don't show up on disk, so we can't include them
+	// when calculating how many "structures" are in gadgetLayout to
+	// ensure that there is only one extra OnDiskStructure at the end
+	numPartsInGadget := 0
+	for _, s := range gadgetLayout.Structure {
+		if s.IsPartition() {
+			numPartsInGadget++
+		}
+
+		// also check for explicit system-data role
+		if s.Role == SystemData {
+			// s can't be implicit system-data since there is an explicit
+			// system-data
+			return false
+		}
+	}
+
+	numPartsOnDisk := len(diskLayout.Structure)
+
+	structEnd := quantity.Offset(s.Size) + s.StartOffset
+	diskEnd := quantity.Offset(uint64(diskLayout.SectorSize) * diskLayout.UsableSectorsEnd)
+
+	return s.Filesystem == "ext4" &&
+		s.Type == "0FC63DAF-8483-4772-8E79-3D69D8477DE4" && // TODO: check hybrid and on MBR/DOS too
+		s.Label == "writable" &&
+		// StructureIndex is 1-based
+		s.StructureIndex == numPartsOnDisk &&
+		numPartsInGadget+1 == numPartsOnDisk &&
+		structEnd == diskEnd
```

Comment:
Question: could this last check byte us later? Isn't it enough to check that this is the last partition?

I'm just wondering if there's a risk that this partition has not been resized to fully use the available space.

Reply:
Hmm, actually looking at this more closely, the thing that does the expanding of the writable partition in uc16/uc18 is the initramfs with this script: https://github.com/snapcore/core-build/blob/master/initramfs/scripts/local-premount/resize and looking at this again, I never realized this before, but actually that script will _always_ resize the writable partition if there is more than 10% free space leftover (and even when doing so actually bricks the device :fearful:). So yes you are correct that there could be a gadget out there which when ubuntu-image prepares the system-data partition implicitly, has less than 10% free space left at the end and thus then doesn't get resized and so this condition would be false when we should still treat that partition as the implicit system-data partition.

For clarity, the sequence here to get that partition in the implicit case is:
1. the gadget snap author doesn't specify the data partition at all in the gadget.yaml
2. In which case ubuntu-image creates it initially with a minimal size with this code: 

https://github.com/canonical/ubuntu-image-legacy/blob/714989a4d0713f0c415f6fa5590a27ba4113c96b/ubuntu_image/common_builder.py#L363-L366

effectively 1.5*(size of files to put in data partition from executing snap prepare-image) + 7 MiB for fs metadata overhead

3. then the initramfs from uc16 and uc18 would dynamically resize the partition on the first boot, which is this code here: 

https://github.com/snapcore/core-build/blob/master/initramfs/scripts/local-premount/resize

Reply:
Just wow :-)